task-maker-async/src/dag.rs

```diff
@@ -0,0 +1,165 @@
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, Hash, PartialEq, Eq)]
+pub enum ExecutionPath {
+    Stdin,
+    Stdout,
+    Stderr,
+    Path(PathBuf),
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, Hash, PartialEq, Eq)]
+pub enum InputFilePermissions {
+    Default,
+    Executable,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, Hash, PartialEq, Eq)]
+pub enum ExecutionFileMode {
+    /// Input file to be read from the store.
+    Input(
+        DataIdentificationHash,
+        VariantIdentificationHash,
+        FileSetFile,
+        InputFilePermissions,
+    ),
+    /// Output file to be written to the store.
+    Output,
+    /// Fifo; executions within the same execution group that share the same Fifo name will share
+    /// the Fifo.
+    Fifo(String),
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Execution {
+    /// Name of the execution, that identifies it within its execution group and is used to refer
+    /// to it.
+    pub name: String,
+    /// Which command to execute.
+    pub command: ExecutionCommand,
+    /// The list of command line arguments.
+    pub args: Vec<String>,
+
+    /// All files (input, output and fifo) that are necessary for running this execution.
+    /// ExecutionPath must not be repeated.
+    pub files: Vec<(ExecutionPath, ExecutionFileMode)>,
+
+    /// Environment variables to set (list of (Key, Value) pairs). The key must not be repeated.
+    pub env: Vec<(String, String)>,
+    /// Environment variables to copy from the sandbox host.
+    pub copy_env: Vec<String>,
+
+    /// Limits on the execution.
+    pub limits: ExecutionLimits,
+
+    /// Other constraints on the execution
+    pub constraints: ExecutionConstraints,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ExecutionGroup {
+    /// A textual description of the group.
+    pub description: String,
+
+    /// The list of executions to run.
+    pub executions: Vec<Execution>,
+
+    /// Setting this field to non-None will cause the execution to skip caching. A different string
+    /// is required for every run in which skipping the cache is desired.
+    pub skip_cache_key: Option<String>,
+
+    /// A priority index for this execution group.
+    pub priority: Priority,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ExecutionDAG {
+    pub execution_groups: Vec<ExecutionGroup>,
+}
+
+impl ExecutionGroup {
+    pub fn get_data_identification_hash(&self) -> DataIdentificationHash {
+        let mut hasher = blake3::Hasher::new();
+        for ex in self.executions.iter() {
+            hasher.update(ex.name.as_bytes());
+            hasher.update(&serialize(&ex.command).unwrap());
+            hasher.update(&serialize(&ex.args).unwrap());
+            for (path, file) in ex.files.iter() {
+                hasher.update(&serialize(path).unwrap());
+                match file {
+                    ExecutionFileMode::Input(data, _, f, perm) => {
```

Comment:
In my opinion replacing `_` with `_variant` makes it clear that the variant is not part of the hash.